activerecord/test/models/parrot.rb

```diff
@@ -18,6 +18,18 @@ def cancel_save_callback_method
   def increment_updated_count
     self.updated_count += 1
   end
+
+  attr_accessor :new_name_input
+  remove_method "new_name_input="
+  def new_name_input=(value)
+    attribute_will_change!(:new_name_input)
+    @new_name_input = value
+  end
+
+  before_update :set_new_name_from_input, if: :new_name_input
+  def set_new_name_from_input
+    self.name = new_name_input
+  end
```

Comment:
You're right, I tried to simplify the test. The changed? method already accounts for changes to virtual attributes, so I used has_changes_to_save? instead.